zqd/ingest/log.go

```diff
@@ -48,20 +50,50 @@ func (rw *recWriter) Write(r *zng.Record) error {
 	return nil
 }
 
-func ingestLogs(ctx context.Context, s *space.Space, paths []string, sortLimit int) error {
-	zr, err := scanner.OpenFiles(resolver.NewContext(), paths...)
-	if err != nil {
+// x509.14:00:00-15:00:00.log.gz (open source zeek)
+// x509_20191101_14:00:00-15:00:00+0000.log.gz (corelight)
+const DefaultJSONPathRegexp = `([a-zA-Z0-9_]+)(?:\.|_\d{8}_)\d\d:\d\d:\d\d\-\d\d:\d\d:\d\d(?:[+\-]\d{4})?\.log(?:$|\.gz)`
+
+func configureJSONTypeReader(ndjr *ndjsonio.Reader, tc ndjsonio.TypeConfig, filename string) error {
+	var path string
+	re := regexp.MustCompile(DefaultJSONPathRegexp)
+	match := re.FindStringSubmatch(filename)
+	if len(match) == 2 {
+		path = match[1]
+	}
+	if err := ndjr.ConfigureTypes(tc, path); err != nil {
 		return err
 	}
-	defer zr.Close()
+	return nil
+}
+
+func ingestLogs(ctx context.Context, s *space.Space, paths []string, tc *ndjsonio.TypeConfig, sortLimit int) error {
+	zctx := resolver.NewContext()
+	var readers []zbuf.Reader
+	for _, path := range paths {
+		sf, err := scanner.OpenFile(zctx, path, "auto")
+		if err != nil {
+			return err
+		}
+		jr, ok := sf.Reader.(*ndjsonio.Reader)
+		if ok && tc != nil {
+			if err = configureJSONTypeReader(jr, *tc, path); err != nil {
+				return err
+			}
+		}
+		readers = append(readers, sf)
+	}
```

Comment:
Nit: We should close the open readers if we encounter an error in this loop.

Reply:
good catch, and sounds like more than a nit to me! fixed.